sycl/include/CL/sycl/ONEAPI/reduction.hpp

```diff
@@ -780,6 +817,217 @@ struct get_reduction_aux_kernel_name_t<sycl::detail::auto_name, Type, B1, B2,
+/// the reduction value hold in \p Reducer is accumulated in those calls.
+template <typename KernelFunc, int Dims, typename ReducerT>
+void reductionLoop(const range<Dims> &Range, ReducerT &Reducer,
+                   const nd_item<1> &NdId, KernelFunc &F) {
+  size_t Start = NdId.get_global_id(0);
+  size_t End = Range.size();
+  size_t Stride = NdId.get_global_range(0);
+  for (size_t I = Start; I < End; I += Stride)
+    F(sycl::detail::getDelinearizedId(Range, I), Reducer);
+}
+
+template <typename KernelName, typename KernelType, int Dims, class Reduction>
+std::enable_if_t<Reduction::has_fast_atomics>
+reduCGFuncImpl(handler &CGH, KernelType KernelFunc, const range<Dims> &Range,
+               const nd_range<1> &NDRange, Reduction &Redu) {
+  auto Out = Redu.getReadWriteAccessorToInitializedMem(CGH);
+  auto GroupSum = Reduction::getReadWriteLocalAcc(1, CGH);
+  using Name =
+      typename get_reduction_main_kernel_name_t<KernelName, KernelType,
+                                                Reduction::is_usm, false>::name;
+  CGH.parallel_for<Name>(NDRange, [=](nd_item<1> NDId) {
+    // Call user's functions. Reducer.MValue gets initialized there.
+    typename Reduction::reducer_type Reducer;
+    reductionLoop(Range, Reducer, NDId, KernelFunc);
+
+    GroupSum[0] = Reducer.getIdentity();
+    sycl::detail::workGroupBarrier();
+    Reducer.template atomic_combine<access::address_space::local_space>(
+        &GroupSum[0]);
+
+    sycl::detail::workGroupBarrier();
+    if (NDId.get_local_id(0) == 0) {
+      Reducer.MValue = GroupSum[0];
+      Reducer.template atomic_combine(Reduction::getOutPointer(Out));
+    }
+  });
+}
+
+template <typename KernelName, typename KernelType, int Dims, class Reduction>
+std::enable_if_t<!Reduction::has_fast_atomics && Reduction::has_fast_reduce>
+reduCGFuncImpl(handler &CGH, KernelType KernelFunc, const range<Dims> &Range,
+               const nd_range<1> &NDRange, Reduction &Redu) {
+  size_t WGSize = NDRange.get_local_range().size();
+  size_t NWorkGroups = NDRange.get_group_range().size();
+
+  bool IsUpdateOfUserVar = !Reduction::is_usm && !Redu.initializeToIdentity();
+  auto PartialSums = Redu.getWriteAccForPartialReds(NWorkGroups, CGH);
+  auto Out =
+      (NWorkGroups == 1) ? PartialSums : Redu.getWriteAccForPartialReds(1, CGH);
+  auto NWorkGroupsFinished =
+      Redu.getReadWriteAccessorToInitializedGroupsCounter(CGH);
+  auto DoReducePartialSumsInLastWG =
+      Reduction::template getReadWriteLocalAcc<int>(1, CGH);
+
+  using Name =
+      typename get_reduction_main_kernel_name_t<KernelName, KernelType,
+                                                Reduction::is_usm, false>::name;
+  CGH.parallel_for<Name>(NDRange, [=](nd_item<1> NDId) {
+    // Call user's functions. Reducer.MValue gets initialized there.
+    typename Reduction::reducer_type Reducer;
+    reductionLoop(Range, Reducer, NDId, KernelFunc);
+
+    DoReducePartialSumsInLastWG[0] = 0;
```

Comment:
Same comment as above.

Reply:
Thank you. Fixed in 808bfa5122a6135f8e6feb45886cc0a59af18528